storage/invoke.go

```diff
@@ -17,28 +17,38 @@ package storage
 import (
 	"context"
 	"errors"
+	"fmt"
 	"io"
 	"net"
+	"net/http"
 	"net/url"
 	"strings"
 
 	"cloud.google.com/go/internal"
+	"cloud.google.com/go/internal/version"
+	sinternal "cloud.google.com/go/storage/internal"
+	"github.com/google/uuid"
 	gax "github.com/googleapis/gax-go/v2"
 	"google.golang.org/api/googleapi"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
 var defaultRetry *retryConfig = &retryConfig{}
+var xGoogDefaultHeader = fmt.Sprintf("gl-go/%s gccl/%s", version.Go(), sinternal.Version)
 
 // run determines whether a retry is necessary based on the config and
 // idempotency information. It then calls the function with or without retries
 // as appropriate, using the configured settings.
-func run(ctx context.Context, call func() error, retry *retryConfig, isIdempotent bool) error {
+func run(ctx context.Context, call func() error, retry *retryConfig, isIdempotent bool, req interface{ Header() http.Header }) error {
```

Comment:
Thinking about how this gets integrated with transport-agnostic clients:

Perhaps `run` accepts not this interface, but rather a function that takes an ID.

```suggestion
func run(ctx context.Context, call func() error, retry *retryConfig, isIdempotent bool, setAttempt func(string))error {
```

Then in transport specific call sites of `run` we just use a closure to define this `func`:

```go
// HTTP
call := c.raw.Buckets.GetIamPolicy(resource)

err = run(ctx, func() error {
  // do stuff
}, c.retry, true, func(attempt string){ setAttemptHeaderHTTP(call, attempt) })
```

```go
// gRPC we set the header using gRPC Metadata on the call context given to the GAPIC
err = run(ctx, func() error {
  // do stuff
}, c.retry, true, func(attempt string){ setAttemptHeaderGRPC(ctx, attempt) })
```

Alternatively, if you want to try to align both HTTP and gRPC on using the call context to communicate Headers/metadata between API layers, maybe there is something that we do in HTTP that pulls the metadata out of the context and adds them to the http.Request.Header 🤷🏻 

I didn't include both the # of attempts and the ID in my example, but you can imagine 😉

Reply:
Even better, we have helpers that create the closure for us:

```go
func httpAttempts(req interface{  Header() http.Header }) func(id string, attempts int) {
  return func(id string, attempts int) {
    setAttemptHeaderHTTP(req, id, attempts)
  }
}

/* inside a method */ 
// HTTP
call := c.raw.Buckets.GetIamPolicy(resource)

err = run(ctx, func() error {
  // do stuff
}, c.retry, true, httpAttempts(call))
```

Reply:
Took your suggestion with the helper func, it works nicely!

Reply:
Great!